pkg/commands/config/canonical_repository.go

```diff
@@ -0,0 +1,72 @@
+package config
+
+import (
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/vmware-tanzu/kpack-cli/pkg/commands"
+	"github.com/vmware-tanzu/kpack-cli/pkg/config"
+	"github.com/vmware-tanzu/kpack-cli/pkg/k8s"
+)
+
+func NewCanonicalRepositoryCommand(clientSetProvider k8s.ClientSetProvider) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "canonical-repository [url]",
+		Short: "Set or Get the canonical repository",
+		Long: `Set or Get the canonical repository 
+
+The canonical repository is the location where imported resources and uploaded local sources are stored. It is required to be configured to use kp.
+
+This data is stored in a config map in the kpack namespace called kp-config. 
+The kp-config config map also contains a service account that contains the secrets required to write to the canonical repository.
+
+If this config map doesn't exist, it will automatically be created by running this command, using the default service account in the kpack namespace as the canonical service account.
+
+
+`,
+		Example: `kp config canonical-repository
+kp config canonical-repository my-registry.com/my-canonical-repo`,
+		Args:         commands.OptionalArgsWithUsage(1),
+		Aliases:      []string{"cr"},
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx := cmd.Context()
+
+			ch, err := commands.NewCommandHelper(cmd)
+			if err != nil {
+				return err
+			}
+
+			cs, err := clientSetProvider.GetClientSet("")
+			if err != nil {
+				return err
+			}
+
+			configHelper := config.NewDefaultConfigHelper(cs)
+
+			if len(args) == 0 {
+				kpConfig, err := configHelper.GetKpConfig(ctx)
+				if err != nil {
+					return errors.Wrap(err, "failed to get canonical repository")
+				}
+
+				repo, err := kpConfig.CanonicalRepository()
+				if err != nil {
+					return errors.Wrap(err, "failed to get canonical repository")
+				}
+
+				return ch.Printlnf("%s", repo)
+			}
+
+			err = configHelper.SetKpConfig(ctx, config.NewKpConfig(args[0], corev1.ObjectReference{}))
```

Comment:
Does this overwrite the service account reference if set?

Reply:
no because it checks to see if the name is empty or not

Reply:
Does that mean you can't zero-out the field via kp?

Reply:
Would distinct set fields methods show a bit clearer intent?